{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import os\n",
    "import json\n",
    "import re\n",
    "\n",
    "from sklearn.ensemble import IsolationForest\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "\n",
    "###########################################\n",
    "#               FUNCTIONS                 #\n",
    "###########################################\n",
    "\n",
    "\n",
    "def clean_reference(ref,outliers):\n",
    "    for i in outliers:\n",
    "        ref = ref.drop(labels=i,axis=1)\n",
    "\n",
    "    return ref\n",
    "\n",
    "def norm_ref(ref):\n",
    "    med = ref.median(axis=0)\n",
    "    norm_ref = ref/med\n",
    "    return norm_ref, med\n",
    "\n",
    "def create_synthetic(norm_ref,med,N):\n",
    "    synt = pd.DataFrame(index=norm_ref.index,columns=range(N))\n",
    "    for j in range(N):\n",
    "        for i in norm_ref.index:\n",
    "            synt[j][i] = np.random.choice(norm_ref.loc[norm_ref.index==i].values.flatten())\n",
    "\n",
    "        synt[j] = synt[j] * np.random.choice(med)\n",
    "    synt.columns = [\"sample_\" + str(i) for i in range(synt.shape[1])]\n",
    "    return synt\n",
    "\n",
    "def add_features(synt,sample,gene,factor,exon=None):\n",
    "    if exon is None:\n",
    "        pattern = gene\n",
    "        tmp = [str(synt.index[i]).split(\"_\")[0] for i in range(synt.shape[0])]\n",
    "        synt.loc[[tmp[i]==pattern for i in range(len(tmp))],sample] = synt.loc[[tmp[i]==pattern for i in range(len(tmp))]][sample] * factor\n",
    "        \n",
    "    if exon is not None:\n",
    "        pattern = gene + \"_\" + exon\n",
    "        tmp = [str(synt.index[i]).split(\"_\")[0] + \"_\" + str(synt.index[i]).split(\"_\")[1] for i in range(synt.shape[0])] \n",
    "        synt.loc[[tmp[i]==pattern for i in range(len(tmp))],sample] = synt.loc[[tmp[i]==pattern for i in range(len(tmp))]][sample] * factor\n",
    "    \n",
    "    res = synt\n",
    "    return res\n",
    "\n",
    "\n",
    "def openJson(path,n):\n",
    "    \"\"\"\n",
    "    Opens json files in path to create a reads matrix\n",
    "    \"\"\"\n",
    "    tmp = os.listdir(path)\n",
    "    tmp = np.array(tmp)[np.array([bool(re.findall(\"depths.json$\",tmp[i])) for i in range(len(tmp))])]\n",
    "    reads = np.zeros((n,len(tmp)))\n",
    "    amplicons = [\"\" for x in range(n)]\n",
    "    q=0\n",
    "    for p in tmp:\n",
    "        with open(path+p) as json_file:\n",
    "            data = json.load(json_file)\n",
    "            for i in range(n):\n",
    "                amplicons[i] = data[i]['name']\n",
    "                for j in data[i]['depths']:\n",
    "                    reads[i,q] = int(data[i]['depths'][j]['min'])\n",
    "        q=q+1\n",
    "    reads = pd.DataFrame(data = reads,index=amplicons)\n",
    "    reads.columns = [i.split('_')[0]+'_'+i.split('_')[1] for i in tmp]\n",
    "    return reads\n",
    "\n",
    "def sumLibraries(reads):\n",
    "    samples = np.unique([i.split('_')[0] for i in reads.columns])\n",
    "    reads_f = np.zeros((reads.shape[0],len(samples)))\n",
    "    q=0\n",
    "    for i in samples:\n",
    "        sub = reads.filter(regex=\"^\"+i)\n",
    "        reads_f[:,q] = sub.sum(axis=1)\n",
    "        q=q+1\n",
    "    reads_f = pd.DataFrame(data = reads_f,index=reads.index)\n",
    "    reads_f.columns = list(samples)\n",
    "    return(reads_f)\n",
    "\n",
    "def correctIndex(reads,correspondance):\n",
    "    l = [\"\" for x in range(len(reads.index))]\n",
    "    q=0\n",
    "    for i in reads.index:\n",
    "        l[q] = i[(len(i)-9):len(i)]\n",
    "        q=q+1\n",
    "    final = reads[[correspondance[\"amplicon\"][0] in l[x] for x in range(len(l))]]\n",
    "    for i in correspondance[\"amplicon\"]:\n",
    "        if i!=correspondance[\"amplicon\"][0]:\n",
    "            final = pd.concat([final,reads[[i in l[x] for x in range(len(l))]]])\n",
    "    final.index = correspondance[\"gene_exon\"] + \"_\" + correspondance[\"amplicon\"]\n",
    "    return final\n",
    "\n",
    "\n",
    "def filterReads(reads,N,output_path):\n",
    "    reads = reads.loc[:,reads.sum(axis=0)>N]\n",
    "    reads = reads.filter(regex=\"^(?!MSI)\",axis=0)\n",
    "    reads = reads.filter(regex=\"^(?!TN)\")\n",
    "    reads = reads.filter(regex=\"^(?!TP)\")\n",
    "    reads = reads.filter(regex=\"^(?!HD)\")\n",
    "    reads = reads.filter(regex=\"^(?!H2)\")\n",
    "    reads.to_csv(output_path, sep=\"\\t\",index=True)\n",
    "    return(reads)\n",
    "\n",
    "\n",
    "def normalizeReads(reads,output_path=None,save=False):\n",
    "    reads_norm=reads/reads.median(axis=0)\n",
    "    reads = np.log(reads+1)\n",
    "    if save==True:\n",
    "        reads_norm.to_csv(output_path, sep=\"\\t\",index=True)\n",
    "    return(reads_norm)\n",
    "\n",
    "\n",
    "def aberrantSamples(reads,conta='auto'):\n",
    "    #reads = reads/np.sum(reads)\n",
    "    \n",
    "    tmp = np.percentile(reads, 99, axis = 0)/np.mean(reads, axis = 0)\n",
    "    random_data = np.array(tmp).reshape(-1,1)\n",
    "    clf = IsolationForest(contamination=conta).fit(random_data)\n",
    "    preds = clf.predict(random_data)\n",
    "    res_amp = np.array(reads.columns)[preds==-1]\n",
    "    \n",
    "    tmp = np.percentile(reads, 1, axis = 0)/np.mean(reads, axis = 0)\n",
    "    random_data = np.array(tmp).reshape(-1,1)\n",
    "    clf = IsolationForest(contamination=conta).fit(random_data)\n",
    "    preds = clf.predict(random_data)\n",
    "    res_del = np.array(reads.columns)[preds==-1]\n",
    "    \n",
    "    res = np.unique(np.concatenate((res_amp,res_del)))\n",
    "    norm = reads.columns[~np.in1d(reads.columns,res)]\n",
    "    \n",
    "    return(res, norm)\n",
    "\n",
    "\n",
    "def aberrantSamples2(reads):\n",
    "    read = reads.astype(\"float\")\n",
    "    tmp = np.percentile(reads, 5, axis = 0)/np.mean(reads, axis = 0)\n",
    "    random_data = np.array(tmp).reshape(-1,1)\n",
    "\n",
    "    clf = IsolationForest(contamination=0.1).fit(random_data)\n",
    "    preds = clf.predict(random_data)\n",
    "    res = np.array(reads.columns)[preds==-1]\n",
    "    return(res)\n",
    "\n",
    "\n",
    "def aberrantAmplicons(reads_norm,abSamples):\n",
    "    for name in res:\n",
    "        random_data = np.array(reads_norm[name]).reshape(-1,1)\n",
    "        clf = IsolationForest(contamination=0.001).fit(np.array(np.mean(reads_norm, axis = 1)).reshape(-1,1))\n",
    "        preds = clf.predict(random_data)\n",
    "        print(name)\n",
    "        print(np.array(reads_norm.index)[preds==-1])\n",
    "\n",
    "def aberrantAmpliconsPerSample(name,reads_norm,conta='auto',verbose=False):\n",
    "    random_data = np.array(reads_norm[name]).reshape(-1,1)\n",
    "    clf = IsolationForest(contamination=conta).fit(np.array(np.mean(reads_norm, axis = 1)).reshape(-1,1))\n",
    "    preds = clf.predict(random_data)\n",
    "    if verbose:\n",
    "        print(name)\n",
    "        print(np.array(reads_norm.index)[preds==-1])\n",
    "    return(np.array(reads_norm.index)[preds==-1])\n",
    "\n",
    "\n",
    "def aberrantAmpliconsPerSample2(name,reads,abSamples,verbose=False):\n",
    "    ab = [i in abSamples for i in reads.columns]\n",
    "    normalReads = reads[np.delete(reads.columns,ab)]\n",
    "    med = np.percentile(normalReads, 99, axis = 1)\n",
    "    reads = (reads.T/med).T\n",
    "    random_data = np.array(reads[name]).reshape(-1,1)\n",
    "    clf = IsolationForest(contamination=0.05).fit(np.array(np.median(reads, axis = 1)).reshape(-1,1))\n",
    "    preds = clf.predict(random_data)\n",
    "    if verbose:\n",
    "        print(name)\n",
    "        print(np.array(reads.index)[preds==-1])\n",
    "    return(np.array(reads.index)[preds==-1])\n",
    "\n",
    "\n",
    "\n",
    "def percentagePerExon(amplified,reads,verbose=False):\n",
    "    genes = [i.split('_')[0] for i in reads.index]\n",
    "    exons = [i.split('_')[1] for i in reads.index]\n",
    "    g_e = [genes[i]+'_'+exons[i] for i in range(len(genes))]\n",
    "    n_ge = np.array([g_e.count(i) for i in np.unique(g_e)])\n",
    "    ag = [i.split('_')[0] for i in amplified]\n",
    "    ae = [i.split('_')[1] for i in amplified]\n",
    "    age = [ag[i]+'_'+ae[i] for i in range(len(amplified))]\n",
    "    f = pd.DataFrame(index=np.unique(age),columns=[\"percentage\"])\n",
    "    f = f.fillna(0)\n",
    "    for i in range(len(np.unique(age))):\n",
    "        f['percentage'][i] = 100*float(age.count(''.join(np.unique(age)[i]))/n_ge[np.unique(g_e)==''.join(np.unique(age)[i])])\n",
    "        if verbose:\n",
    "            if f['percentage'][i]>50:\n",
    "                print(np.unique(age)[i] + \": \" + str(round(f['percentage'][i]))+'%'+' des amplicons de l\\'exon sont aberrants')\n",
    "    return(f)\n",
    "\n",
    "def percentagePerGene(amplified,reads,verbose=False):\n",
    "    genes = [i.split('_')[0] for i in reads.index]\n",
    "    ag = [i.split('_')[0] for i in amplified]\n",
    "    n_g = np.array([genes.count(i) for i in np.unique(genes)])\n",
    "    f = pd.DataFrame(index=np.unique(ag),columns=[\"percentage\"])\n",
    "    f = f.fillna(0)\n",
    "    for i in range(len(np.unique(ag))):\n",
    "        f['percentage'][i] = 100*float(ag.count(''.join(np.unique(ag)[i]))/n_g[np.unique(genes)==''.join(np.unique(ag)[i])])\n",
    "        if verbose:\n",
    "            if f['percentage'][i]>50:\n",
    "                print(np.unique(ag)[i] + \": \" + str(round(f['percentage'][i]))+'%'+' des amplicons du gene sont aberrants')\n",
    "    return(f)\n",
    "\n",
    "def amplifEvalGene(reads,abSamples,gene,sample):\n",
    "    reads_m = reads/reads.median(axis=0)\n",
    "    sub = reads_m\n",
    "    for i in abSamples:\n",
    "        sub = sub.drop(labels=i,axis=1)\n",
    "    reads_m = reads_m.filter(regex=\"^\"+gene,axis=0)\n",
    "    reads_m = reads_m[sample]   \n",
    "    val = np.mean(reads_m)/np.mean(sub.mean())\n",
    "    if val==np.inf:\n",
    "        val = 100\n",
    "    return val\n",
    "\n",
    "def scoreAmplif(k,n,N):\n",
    "    p = n/N\n",
    "    x = np.log(1/((p**k)*(1-p)**(n-k)))*(k/n)\n",
    "    # score = 1/(1+np.exp(-x))\n",
    "    score = x/390 + 190/390\n",
    "    \n",
    "    return x\n",
    "\n",
    "def aberrantAmpliconsFinal(reads, reads_norm, abSamples,abSamples2,run,threshold):\n",
    "    f = pd.DataFrame(columns=[\"run\",\"name\",\"gene\",\"amplif\",\"score\"])\n",
    "        \n",
    "    q=0 \n",
    "    for name in abSamples2:\n",
    "        #abAmp = aberrantAmpliconsPerSample2(name,reads_norm,abSamples,verbose=False)\n",
    "        abAmp = aberrantAmpliconsPerSample(name,reads_norm,verbose=False)\n",
    "        if abAmp.shape!=(0,):\n",
    "            genes = np.unique([i.split('_')[0] for i in abAmp])\n",
    "            for gene in genes:\n",
    "                r = re.compile(gene)\n",
    "                abEx = list(filter(r.match, abAmp))\n",
    "                exons1 = [i.split('_')[0]+\"_\"+i.split('_')[1] for i in abEx]\n",
    "                tmp = reads.filter(regex=\"^\"+gene,axis=0)\n",
    "                exons2 = [i.split('_')[0]+\"_\"+i.split('_')[1] for i in tmp.index]\n",
    "                \n",
    "                score = scoreAmplif(len(abEx),tmp.shape[0],reads.shape[0])\n",
    "                \n",
    "                amplif = amplifEvalGene(reads, abSamples, gene, name)\n",
    "\n",
    "                if score>threshold:\n",
    "                    if amplif>1:\n",
    "                        f.loc[q] = [run,name,gene,amplif,score]\n",
    "                        q=q+1\n",
    "                    #if amplif<1:\n",
    "                    #    f.loc[q] = [run,name,gene,amplif,score]\n",
    "                    #    q=q+1\n",
    "\n",
    "    return(f)\n",
    "\n",
    "\n",
    "def aberrantAmpliconsFinal2(reads, reads_norm, abSamples,abSamples2,run,threshold):\n",
    "    f = pd.DataFrame(columns=[\"run\",\"name\",\"gene\",\"amplif\",\"score\"])\n",
    "        \n",
    "    q=0 \n",
    "    for name in abSamples2:\n",
    "        #abAmp = aberrantAmpliconsPerSample2(name,reads_norm,abSamples,verbose=False)\n",
    "        abAmp = aberrantAmpliconsPerSample(name,reads_norm,verbose=False)\n",
    "        if abAmp.shape!=(0,):\n",
    "            genes = abAmp\n",
    "            for gene in genes:\n",
    "                r = re.compile(gene)\n",
    "                abEx = list(filter(r.match, abAmp))\n",
    "                #print(abEx)\n",
    "                tmp = reads.filter(regex=\"^\"+gene,axis=0)                \n",
    "                score = scoreAmplif(len(abEx),tmp.shape[0],reads.shape[0])\n",
    "                \n",
    "                amplif = amplifEvalGene(reads, abSamples, gene, name)\n",
    "\n",
    "                if score>threshold:\n",
    "                    if amplif>1:\n",
    "                        f.loc[q] = [run,name,gene,amplif,score]\n",
    "                        q=q+1\n",
    "                    #if amplif<1:\n",
    "                    #    f.loc[q] = [run,name,gene,amplif,score]\n",
    "                    #    q=q+1\n",
    "\n",
    "    return(f)\n",
    "\n",
    "\n",
    "\n",
    "def aberrantTargetsCapture(abSamples, normalSamples, reads_norm, conta=None, lower=-0.5, upper=0.5, verbose=True, output_path=False):\n",
    "    if conta == None:\n",
    "        conta = 1/reads_norm.shape[1]\n",
    "    f = pd.DataFrame(columns=[\"name\",\"loc\",\"amp\"])\n",
    "    \n",
    "    norm = np.mean(reads_norm[normalSamples], axis = 1)\n",
    "\n",
    "    q=0\n",
    "    for i in tqdm(reads_norm.columns):            \n",
    "        x = np.array(reads_norm[i]/norm)\n",
    "\n",
    "        if sum(reads_norm[i]/norm>1.5)>1:\n",
    "            #print('duplication: '+i+' '+str(sum(reads_norm[i]/norm>1.4)))\n",
    "            gg=1\n",
    "        if sum(reads_norm[i]/norm<0.6)>1:\n",
    "            #print('deletion:'+i+' '+str(sum(reads_norm[i]/norm<0.6)))\n",
    "            gg=1        \n",
    "        \n",
    "        \n",
    "        if gg==1:\n",
    "            random_data = x.reshape(-1, 1)\n",
    "            clf = IsolationForest(contamination=conta).fit(random_data)\n",
    "            preds = clf.predict(random_data)\n",
    "\n",
    "            det = np.array(reads_norm.index)[preds==-1]\n",
    "\n",
    "            for j in det:\n",
    "                amp = float(x[reads_norm.index==j])\n",
    "                if amp<lower:\n",
    "                    f.loc[q] = [i,j,amp]\n",
    "                    q=q+1\n",
    "                if amp>upper:\n",
    "                    f.loc[q] = [i,j,amp]\n",
    "                    q=q+1\n",
    "        gg=0\n",
    "    \n",
    "    #for i in tqdm(reads_norm.index):\n",
    "#\n",
    "    #    x = np.array(np.log2(reads_norm.loc[i]/float(norm[reads_norm.index==i])))\n",
    "    #    #pd.DataFrame(x).to_csv(\"/Users/admin/Documents/CNV/testttt.tsv\",sep=\"\\t\")\n",
    "    #    random_data = x.reshape(-1, 1)\n",
    "    #    clf = IsolationForest(contamination=conta).fit(random_data)\n",
    "    #    preds = clf.predict(random_data)\n",
    "#\n",
    "    #    det = np.array(final_norm.columns)[preds==-1]\n",
    "    #    ndet = np.array(final_norm.columns)[preds==1]\n",
    "#\n",
    "    #    for j in det:\n",
    "    #        amp = x[final_norm.columns==j]\n",
    "    #        if amp<lower:\n",
    "    #            f.loc[q] = [j,i,amp]\n",
    "    #            q=q+1\n",
    "    #        if amp>upper:\n",
    "    #            f.loc[q] = [j,i,amp]\n",
    "    #            q=q+1\n",
    "                \n",
    "    if verbose:\n",
    "        print(str(f.shape[0])+\" aberrant targets detected in \"+str(len(np.unique(f['name'])))+\" samples\")\n",
    "    return f\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0543e73cf80e4ebd9630164a3e31fc17",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=96.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "16234 aberrant targets detected in 69 samples\n"
     ]
    }
   ],
   "source": [
    "pd.set_option('display.max_rows', 2000)\n",
    "output_path = \"/Users/admin/Documents/CNV/\"\n",
    "reads = pd.read_csv(\"/Users/admin/Documents/CNV/final_ICR96_good.txt\",sep=\"\\t\",index_col=0)\n",
    "\n",
    "abSamples, normalSamples = aberrantSamples(reads,conta=0.5)\n",
    "#pd.DataFrame(normalSamples).to_csv(\"/Users/admin/Documents/CNV/normal_samples_icr.tsv\",sep=\"\\t\")\n",
    "\n",
    "final_norm = normalizeReads(reads,output_path,save=False)\n",
    "\n",
    "f = aberrantTargetsCapture(abSamples, normalSamples, final_norm,conta='auto',upper=0,lower=0)\n",
    "\n",
    "f.to_csv(\"/Users/admin/Documents/CNV/icr_res/res_capture_ICR96_test.txt\", sep=\"\\t\",index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "627a009e668c4887869a76f5e228861f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=25.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "from tqdm.notebook import tqdm\n",
    "\n",
    "tp=[]\n",
    "fp=[]\n",
    "tn=[]\n",
    "fn=[]\n",
    "fac=[]\n",
    "for k in tqdm(range(1,26)):\n",
    "    for p in (range(100)):\n",
    "        synt = pd.read_csv(\"/Users/admin/Documents/CNV/synthetic_raw_85.tsv\",sep=\"\\t\",index_col=0)\n",
    "        synt = synt[np.random.choice(synt.columns,50,replace=False)]\n",
    "        col = np.random.choice(synt.columns,size=k,replace=False)\n",
    "        normalSamples = synt.columns[~np.in1d(synt.columns,col)]\n",
    "\n",
    "        row = np.random.choice(synt.index,size=10)\n",
    "        #row = pd.read_csv(\"/Users/admin/Documents/CNV/article/figs/data_fig3/rows.tsv\",sep=\"\\t\")\n",
    "        #row = np.array(row)\n",
    "\n",
    "        neg = synt.columns[~np.in1d(synt.columns,col)]\n",
    "        pos = col\n",
    "        \n",
    "        for j in col:\n",
    "            factor = np.random.choice(range(2,7))\n",
    "            fac.extend([factor])\n",
    "            for i in row:\n",
    "                synt.at[i,j] = synt.at[i,j]*factor\n",
    "\n",
    "        synt_norm = normalizeReads(synt)\n",
    "\n",
    "        conta = 'auto'#1/synt_norm.shape[0]\n",
    "        \n",
    "        det, ndet = aberrantSamples(synt,conta = conta)\n",
    "        \n",
    "        #norm = np.mean(synt_norm[normalSamples], axis = 1)\n",
    "\n",
    "        #x = np.array(synt_norm[col]/norm[0])     \n",
    "\n",
    "        #random_data = x.reshape(-1, 1)\n",
    "        #clf = IsolationForest(contamination=conta).fit(random_data)\n",
    "        #preds = clf.predict(random_data)\n",
    "\n",
    "        #det = np.array(synt_norm.index)[preds==-1]\n",
    "        #ndet = np.array(synt_norm.index)[preds==1]\n",
    "\n",
    "        tp.extend([sum(np.in1d(det,pos))])\n",
    "        fp.extend([sum(np.in1d(det,neg))])\n",
    "        tn.extend([sum(np.in1d(ndet,neg))])\n",
    "        fn.extend([sum(np.in1d(ndet,pos))])\n",
    "\n",
    "pd.DataFrame(tp).to_csv(\"/Users/admin/Documents/CNV/article/figs/data_fig2/tp_kN.tsv\", sep=\"\\t\",index=False)\n",
    "pd.DataFrame(fp).to_csv(\"/Users/admin/Documents/CNV/article/figs/data_fig2/fp_kN.tsv\", sep=\"\\t\",index=False)\n",
    "pd.DataFrame(fn).to_csv(\"/Users/admin/Documents/CNV/article/figs/data_fig2/fn_kN.tsv\", sep=\"\\t\",index=False)\n",
    "pd.DataFrame(tn).to_csv(\"/Users/admin/Documents/CNV/article/figs/data_fig2/tn_kN.tsv\", sep=\"\\t\",index=False)\n",
    "pd.DataFrame(fac).to_csv(\"/Users/admin/Documents/CNV/article/figs/data_fig2/factors_kN.tsv\", sep=\"\\t\",index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'sample_32'"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "j"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c757da2bcbc14d55a88bd91424293cec",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=10.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['JAK1_E20_STA100335']\n",
      "['JAK1_E18_STA100342']\n",
      "['JAK1_E14_STA100350']\n",
      "['JAK1_E13_STA100352']\n",
      "['JAK1_E5_STA100377']\n",
      "['JAK1_E4_STA100381']\n",
      "['DDR2_E5_STA100388']\n",
      "['DDR2_E5_STA100390']\n",
      "['DDR2_E9_STA100401']\n",
      "['DDR2_E10_STA100406']\n",
      "['DDR2_E10_STA100407']\n",
      "['DDR2_E10_STA100408']\n",
      "['DDR2_E12_STA100412']\n",
      "['DDR2_E13_STA100415']\n",
      "['DDR2_E14_STA100416']\n",
      "['DDR2_E14_STA101731']\n",
      "['DDR2_E16_STA100423']\n",
      "['H3F3A_E2_STA100009']\n",
      "['H3F3A_E2_STA100011']\n",
      "['ALK_E29_STA101368']\n",
      "['ALK_E29_STA100107']\n",
      "['ALK_E29_STA100109']\n",
      "['ALK_E26_STA100117']\n",
      "['ALK_E22_STA100125']\n",
      "['ALK_E20_STA100130']\n",
      "['ALK_E17_STA101372']\n",
      "['ALK_E14_STA101375']\n",
      "['IDH1_E4_STA100134']\n",
      "['ERBB4_E11_STA101738']\n",
      "['VHL_E3_STA101156']\n",
      "['CTNNB1_E3_STA100162']\n",
      "['CTNNB1_E3_STA100164']\n",
      "['MITF_E10_STA100166']\n",
      "['FOXL2_E1_STA103226']\n",
      "['PIK3CA_E2_STA101624']\n",
      "['PIK3CA_E4_STA101651']\n",
      "['PIK3CA_E10_STA100190']\n",
      "['PIK3CA_E18_STA101403']\n",
      "['FGFR3_E3_STA101676']\n",
      "['FGFR3_E7_STA101724']\n",
      "['PDGFRA_E4_STA101422']\n",
      "['PDGFRA_E8_STA101426']\n",
      "['PDGFRA_E11_STA101429']\n",
      "['PDGFRA_E12_STA100209']\n",
      "['PDGFRA_E14_STA100210']\n",
      "['PDGFRA_E17_STA101433']\n",
      "['KIT_E1_STA101439']\n",
      "['KIT_E2_STA101440']\n",
      "['KIT_E3_STA101441']\n",
      "['KIT_E4_STA101442']\n",
      "['KIT_E5_STA101443']\n",
      "['KIT_E6_STA101444']\n",
      "['KIT_E7_STA101445']\n",
      "['KIT_E8_STA100215']\n",
      "['KIT_E8_STA100216']\n",
      "['JAK1_E24_STA100324']\n",
      "['JAK1_E23_STA100325']\n",
      "['JAK1_E23_STA100326']\n",
      "['JAK1_E21_STA100333']\n",
      "['JAK1_E20_STA100335']\n",
      "['JAK1_E20_STA100336']\n",
      "['JAK1_E18_STA100342']\n",
      "['JAK1_E14_STA100351']\n",
      "['JAK1_E13_STA100353']\n",
      "['JAK1_E12_STA100355']\n",
      "['JAK1_E10_STA100360']\n",
      "['JAK1_E9_STA100362']\n",
      "['JAK1_E6_STA100375']\n",
      "['JAK1_E5_STA100377']\n",
      "['JAK1_E5_STA100378']\n",
      "['JAK1_E4_STA100380']\n",
      "['DDR2_E6_STA100391']\n",
      "['DDR2_E8_STA100399']\n",
      "['DDR2_E9_STA100403']\n",
      "['DDR2_E10_STA100407']\n",
      "['DDR2_E10_STA100408']\n",
      "['DDR2_E13_STA100413']\n",
      "['DDR2_E15_STA100420']\n",
      "['DDR2_E16_STA100422']\n",
      "['ALK_E25_STA100119']\n",
      "['ALK_E23_STA100124']\n",
      "['ALK_E22_STA100125']\n",
      "['ALK_E20_STA101611']\n",
      "['ALK_E17_STA101372']\n",
      "['ALK_E6_STA101383']\n",
      "['ALK_E1_STA102007']\n",
      "['ERBB4_E12_STA100136']\n",
      "['ERBB4_E10_STA100140']\n",
      "['VHL_E3_STA101156']\n",
      "['CTNNB1_E3_STA100162']\n",
      "['PIK3CA_E6_STA101393']\n",
      "['PIK3CA_E8_STA101395']\n",
      "['PIK3CA_E16_STA101401']\n",
      "['PIK3CA_E21_STA101406']\n",
      "['PIK3CA_E21_STA100195']\n",
      "['PIK3CA_E21_STA100196']\n",
      "['FGFR3_E3_STA101676']\n",
      "['FGFR3_E9_STA100201']\n",
      "['PDGFRA_E5_STA101423']\n",
      "['PDGFRA_E14_STA100211']\n",
      "['PDGFRA_E21_STA101436']\n",
      "['KIT_E1_STA101439']\n",
      "['KIT_E2_STA101440']\n",
      "['KIT_E3_STA101441']\n",
      "['KIT_E4_STA101442']\n",
      "['KIT_E5_STA101443']\n",
      "['KIT_E6_STA101444']\n",
      "['KIT_E7_STA101445']\n",
      "['KIT_E8_STA100215']\n",
      "['KIT_E8_STA100216']\n",
      "['JAK1_E24_STA100324']\n",
      "['JAK1_E21_STA100333']\n",
      "['JAK1_E19_STA100338']\n",
      "['JAK1_E17_STA100343']\n",
      "['JAK1_E13_STA100353']\n",
      "['JAK1_E8_STA100367']\n",
      "['JAK1_E7_STA100370']\n",
      "['JAK1_E2_STA100386']\n",
      "['NRAS_E3_STA101617']\n",
      "['DDR2_E6_STA100394']\n",
      "['DDR2_E7_STA100396']\n",
      "['DDR2_E9_STA100403']\n",
      "['DDR2_E10_STA100405']\n",
      "['DDR2_E10_STA100409']\n",
      "['DDR2_E11_STA100410']\n",
      "['DDR2_E13_STA100414']\n",
      "['DDR2_E14_STA100416']\n",
      "['DDR2_E14_STA101731']\n",
      "['DDR2_E14_STA100419']\n",
      "['DDR2_E16_STA100422']\n",
      "['DDR2_E17_STA100425']\n",
      "['DDR2_E17_STA100427']\n",
      "['DDR2_E19_STA100433']\n",
      "['ALK_E29_STA100108']\n",
      "['ALK_E29_STA101369']\n",
      "['ALK_E27_STA100115']\n",
      "['ALK_E20_STA100128']\n",
      "['ALK_E19_STA101370']\n",
      "['ALK_E10_STA101379']\n",
      "['ALK_E9_STA101380']\n",
      "['VHL_E3_STA101154']\n",
      "['VHL_E3_STA101155']\n",
      "['CTNNB1_E3_STA100160']\n",
      "['MITF_E10_STA100167']\n",
      "['PIK3CA_E3_STA100185']\n",
      "['PIK3CA_E8_STA101395']\n",
      "['PIK3CA_E12_STA101397']\n",
      "['PIK3CA_E17_STA101402']\n",
      "['PIK3CA_E21_STA100194']\n",
      "['PIK3CA_E21_STA100195']\n",
      "['PIK3CA_E21_STA100196']\n",
      "['FGFR3_E3_STA101676']\n",
      "['FGFR3_E5_STA101410']\n",
      "['FGFR3_E7_STA103215']\n",
      "['FGFR3_E10_STA101413']\n",
      "['PDGFRA_E7_STA101425']\n",
      "['PDGFRA_E12_STA100207']\n",
      "['PDGFRA_E13_STA101430']\n",
      "['PDGFRA_E14_STA100210']\n",
      "['PDGFRA_E16_STA101432']\n",
      "['PDGFRA_E23_STA101438']\n",
      "['KIT_E1_STA101439']\n",
      "['KIT_E2_STA101440']\n",
      "['KIT_E3_STA101441']\n",
      "['KIT_E4_STA101442']\n",
      "['KIT_E5_STA101443']\n",
      "['KIT_E6_STA101444']\n",
      "['KIT_E7_STA101445']\n",
      "['KIT_E8_STA100215']\n",
      "['KIT_E8_STA100216']\n",
      "['JAK1_E24_STA100323']\n",
      "['JAK1_E21_STA100331']\n",
      "['JAK1_E20_STA100335']\n",
      "['JAK1_E19_STA100338']\n",
      "['JAK1_E19_STA100339']\n",
      "['JAK1_E14_STA100351']\n",
      "['JAK1_E6_STA100374']\n",
      "['NRAS_E4_STA100002']\n",
      "['NRAS_E3_STA101617']\n",
      "['DDR2_E8_STA100398']\n",
      "['DDR2_E9_STA100401']\n",
      "['DDR2_E10_STA100405']\n",
      "['DDR2_E14_STA100416']\n",
      "['DDR2_E14_STA100418']\n",
      "['DDR2_E14_STA100419']\n",
      "['DDR2_E16_STA100423']\n",
      "['DDR2_E17_STA100425']\n",
      "['DDR2_E18_STA100429']\n",
      "['ALK_E29_STA100106']\n",
      "['ALK_E29_STA100108']\n",
      "['ALK_E29_STA101369']\n",
      "['ALK_E25_STA100119']\n",
      "['ALK_E22_STA100125']\n",
      "['ALK_E20_STA101611']\n",
      "['IDH1_E4_STA100133']\n",
      "['MITF_E10_STA100167']\n",
      "['FOXL2_E1_STA103227']\n",
      "['PIK3CA_E3_STA100187']\n",
      "['PIK3CA_E6_STA101393']\n",
      "['PIK3CA_E11_STA100191']\n",
      "['PIK3CA_E16_STA101401']\n",
      "['PIK3CA_E17_STA101402']\n",
      "['PIK3CA_E19_STA101404']\n",
      "['PIK3CA_E21_STA101406']\n",
      "['PIK3CA_E21_STA100194']\n",
      "['PIK3CA_E21_STA100196']\n",
      "['FGFR3_E17_STA101418']\n",
      "['PDGFRA_E8_STA101426']\n",
      "['PDGFRA_E11_STA101429']\n",
      "['PDGFRA_E14_STA100210']\n",
      "['PDGFRA_E18_STA100213']\n",
      "['PDGFRA_E18_STA100214']\n",
      "['PDGFRA_E23_STA101438']\n",
      "['KIT_E1_STA101439']\n",
      "['KIT_E2_STA101440']\n",
      "['KIT_E3_STA101441']\n",
      "['KIT_E4_STA101442']\n",
      "['KIT_E5_STA101443']\n",
      "['KIT_E6_STA101444']\n",
      "['KIT_E7_STA101445']\n",
      "['KIT_E8_STA100215']\n",
      "['KIT_E8_STA100216']\n",
      "['JAK1_E25_STA100320']\n",
      "['JAK1_E20_STA100335']\n",
      "['JAK1_E17_STA100344']\n",
      "['JAK1_E14_STA100351']\n",
      "['JAK1_E7_STA100372']\n",
      "['JAK1_E6_STA100375']\n",
      "['JAK1_E5_STA100376']\n",
      "['JAK1_E5_STA101781']\n",
      "['NRAS_E4_STA100002']\n",
      "['DDR2_E9_STA100401']\n",
      "['DDR2_E9_STA100403']\n",
      "['DDR2_E12_STA100412']\n",
      "['DDR2_E15_STA100420']\n",
      "['DDR2_E16_STA100422']\n",
      "['DDR2_E17_STA100427']\n",
      "['ALK_E29_STA100107']\n",
      "['ALK_E29_STA100109']\n",
      "['ALK_E29_STA101369']\n",
      "['ALK_E27_STA101615']\n",
      "['ALK_E20_STA101611']\n",
      "['ALK_E20_STA100129']\n",
      "['ALK_E13_STA101376']\n",
      "['ALK_E11_STA101378']\n",
      "['ALK_E10_STA101379']\n",
      "['ALK_E7_STA101382']\n",
      "['CTNNB1_E3_STA100161']\n",
      "['CTNNB1_E3_STA100164']\n",
      "['MITF_E10_STA100165']\n",
      "['FOXL2_E1_STA103226']\n",
      "['PIK3CA_E4_STA101651']\n",
      "['PIK3CA_E6_STA101393']\n",
      "['PIK3CA_E10_STA100188']\n",
      "['PIK3CA_E21_STA100194']\n",
      "['FGFR3_E3_STA101676']\n",
      "['FGFR3_E5_STA101410']\n",
      "['FGFR3_E16_STA101722']\n",
      "['PDGFRA_E6_STA101424']\n",
      "['PDGFRA_E7_STA101425']\n",
      "['PDGFRA_E14_STA100211']\n",
      "['PDGFRA_E14_STA100212']\n",
      "['PDGFRA_E21_STA101436']\n",
      "['KIT_E1_STA101439']\n",
      "['KIT_E4_STA101442']\n",
      "['KIT_E6_STA101444']\n",
      "['KIT_E7_STA101445']\n",
      "['KIT_E8_STA100215']\n",
      "['KIT_E8_STA100216']\n",
      "['JAK1_E22_STA100328']\n",
      "['JAK1_E17_STA100344']\n",
      "['JAK1_E10_STA100360']\n",
      "['JAK1_E5_STA100377']\n",
      "['JAK1_E4_STA100381']\n",
      "['NRAS_E2_STA100007']\n",
      "['DDR2_E5_STA100388']\n",
      "['DDR2_E9_STA100401']\n",
      "['DDR2_E13_STA100413']\n",
      "['DDR2_E13_STA100414']\n",
      "['DDR2_E14_STA100416']\n",
      "['DDR2_E14_STA101731']\n",
      "['DDR2_E16_STA100423']\n",
      "['DDR2_E19_STA100433']\n",
      "['ALK_E29_STA100107']\n",
      "['ALK_E24_STA100122']\n",
      "['ALK_E23_STA100123']\n",
      "['IDH1_E4_STA100133']\n",
      "['ERBB4_E12_STA100136']\n",
      "['CTNNB1_E3_STA100162']\n",
      "['PIK3CA_E3_STA100185']\n",
      "['PIK3CA_E10_STA100188']\n",
      "['PIK3CA_E10_STA101626']\n",
      "['PIK3CA_E11_STA100192']\n",
      "['FGFR3_E5_STA101410']\n",
      "['FGFR3_E9_STA101705']\n",
      "['FGFR3_E9_STA101708']\n",
      "['PDGFRA_E12_STA101616']\n",
      "['KIT_E2_STA101440']\n",
      "['KIT_E4_STA101442']\n",
      "['KIT_E5_STA101443']\n",
      "['KIT_E7_STA101445']\n",
      "['KIT_E8_STA100215']\n",
      "['KIT_E8_STA100216']\n",
      "['JAK1_E14_STA100350']\n",
      "['JAK1_E13_STA100353']\n",
      "['JAK1_E10_STA100359']\n",
      "['JAK1_E10_STA100360']\n",
      "['JAK1_E6_STA100375']\n",
      "['JAK1_E4_STA100382']\n",
      "['DDR2_E5_STA100389']\n",
      "['DDR2_E6_STA100391']\n",
      "['DDR2_E9_STA100401']\n",
      "['DDR2_E9_STA100402']\n",
      "['DDR2_E13_STA100413']\n",
      "['DDR2_E14_STA100419']\n",
      "['DDR2_E15_STA100420']\n",
      "['DDR2_E15_STA100421']\n",
      "['DDR2_E16_STA100422']\n",
      "['ALK_E29_STA100103']\n",
      "['ALK_E29_STA100109']\n",
      "['ALK_E28_STA100111']\n",
      "['ALK_E27_STA100114']\n",
      "['ALK_E21_STA100126']\n",
      "['ALK_E20_STA100128']\n",
      "['ALK_E10_STA101379']\n",
      "['ALK_E8_STA101381']\n",
      "['ALK_E4_STA101385']\n",
      "['PIK3CA_E2_STA100182']\n",
      "['PIK3CA_E4_STA101651']\n",
      "['PIK3CA_E12_STA101397']\n",
      "['PIK3CA_E14_STA101649']\n",
      "['FGFR3_E5_STA101410']\n",
      "['FGFR3_E9_STA100201']\n",
      "['FGFR3_E16_STA101722']\n",
      "['PDGFRA_E2_STA101420']\n",
      "['PDGFRA_E6_STA101424']\n",
      "['PDGFRA_E8_STA101426']\n",
      "['PDGFRA_E9_STA101427']\n",
      "['PDGFRA_E12_STA100209']\n",
      "['PDGFRA_E14_STA100211']\n",
      "['PDGFRA_E16_STA101432']\n",
      "['PDGFRA_E17_STA101433']\n",
      "['PDGFRA_E21_STA101436']\n",
      "['PDGFRA_E23_STA101438']\n",
      "['KIT_E1_STA101439']\n",
      "['KIT_E2_STA101440']\n",
      "['KIT_E3_STA101441']\n",
      "['KIT_E4_STA101442']\n",
      "['KIT_E5_STA101443']\n",
      "['KIT_E8_STA100216']\n",
      "['JAK1_E19_STA100339']\n",
      "['JAK1_E14_STA100351']\n",
      "['JAK1_E11_STA100356']\n",
      "['JAK1_E6_STA100375']\n",
      "['JAK1_E5_STA100377']\n",
      "['JAK1_E5_STA100378']\n",
      "['JAK1_E2_STA100386']\n",
      "['DDR2_E5_STA100388']\n",
      "['DDR2_E7_STA100395']\n",
      "['DDR2_E9_STA100401']\n",
      "['DDR2_E11_STA100410']\n",
      "['DDR2_E13_STA100413']\n",
      "['DDR2_E14_STA101731']\n",
      "['DDR2_E15_STA100421']\n",
      "['DDR2_E16_STA100424']\n",
      "['DDR2_E17_STA100425']\n",
      "['DDR2_E18_STA100429']\n",
      "['ALK_E20_STA100130']\n",
      "['ALK_E13_STA101376']\n",
      "['ALK_E12_STA101377']\n",
      "['ALK_E6_STA101383']\n",
      "['ERBB4_E11_STA100138']\n",
      "['FOXL2_E1_STA103227']\n",
      "['PIK3CA_E2_STA101624']\n",
      "['PIK3CA_E3_STA100185']\n",
      "['PIK3CA_E10_STA100188']\n",
      "['PIK3CA_E21_STA100196']\n",
      "['FGFR3_E5_STA101410']\n",
      "['PDGFRA_E12_STA100207']\n",
      "['PDGFRA_E12_STA101616']\n",
      "['PDGFRA_E16_STA101432']\n",
      "['PDGFRA_E18_STA100213']\n",
      "['PDGFRA_E22_STA101437']\n",
      "['PDGFRA_E23_STA101438']\n",
      "['KIT_E2_STA101440']\n",
      "['KIT_E4_STA101442']\n",
      "['KIT_E8_STA100216']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['JAK1_E22_STA100329']\n",
      "['JAK1_E20_STA100334']\n",
      "['JAK1_E19_STA100338']\n",
      "['JAK1_E13_STA100352']\n",
      "['JAK1_E9_STA100362']\n",
      "['JAK1_E6_STA100375']\n",
      "['JAK1_E5_STA100376']\n",
      "['NRAS_E2_STA100007']\n",
      "['DDR2_E9_STA100401']\n",
      "['DDR2_E12_STA100412']\n",
      "['DDR2_E13_STA100414']\n",
      "['DDR2_E13_STA100415']\n",
      "['DDR2_E14_STA101731']\n",
      "['DDR2_E16_STA100423']\n",
      "['DDR2_E17_STA100426']\n",
      "['DDR2_E17_STA100427']\n",
      "['H3F3A_E2_STA101614']\n",
      "['ALK_E29_STA100108']\n",
      "['ALK_E29_STA100109']\n",
      "['ALK_E20_STA101611']\n",
      "['ALK_E20_STA100128']\n",
      "['ALK_E20_STA100130']\n",
      "['IDH1_E4_STA100134']\n",
      "['FOXL2_E1_STA103228']\n",
      "['PIK3CA_E2_STA101624']\n",
      "['PIK3CA_E21_STA100195']\n",
      "['PIK3CA_E21_STA100196']\n",
      "['FGFR3_E16_STA101722']\n",
      "['PDGFRA_E11_STA101429']\n",
      "['PDGFRA_E15_STA101431']\n",
      "['PDGFRA_E18_STA100214']\n",
      "['PDGFRA_E22_STA101437']\n",
      "['KIT_E2_STA101440']\n",
      "['KIT_E6_STA101444']\n",
      "['JAK1_E25_STA100320']\n",
      "['JAK1_E25_STA100321']\n",
      "['JAK1_E24_STA100323']\n",
      "['JAK1_E24_STA100324']\n",
      "['JAK1_E22_STA100327']\n",
      "['JAK1_E16_STA100346']\n",
      "['JAK1_E14_STA100351']\n",
      "['JAK1_E13_STA100353']\n",
      "['JAK1_E11_STA100358']\n",
      "['JAK1_E10_STA100359']\n",
      "['JAK1_E9_STA100362']\n",
      "['JAK1_E6_STA100374']\n",
      "['JAK1_E5_STA100376']\n",
      "['DDR2_E6_STA100391']\n",
      "['DDR2_E7_STA100396']\n",
      "['DDR2_E8_STA100399']\n",
      "['DDR2_E9_STA100401']\n",
      "['DDR2_E14_STA101731']\n",
      "['DDR2_E14_STA100418']\n",
      "['DDR2_E15_STA100420']\n",
      "['DDR2_E16_STA100422']\n",
      "['DDR2_E19_STA100431']\n",
      "['DDR2_E19_STA100432']\n",
      "['ALK_E23_STA100124']\n",
      "['ALK_E19_STA101370']\n",
      "['ALK_E13_STA101376']\n",
      "['ALK_E12_STA101377']\n",
      "['ALK_E11_STA101378']\n",
      "['ALK_E6_STA101383']\n",
      "['IDH1_E4_STA100134']\n",
      "['VHL_E3_STA101154']\n",
      "['CTNNB1_E3_STA100160']\n",
      "['FOXL2_E1_STA103228']\n",
      "['FOXL2_E1_STA103226']\n",
      "['FOXL2_E1_STA103227']\n",
      "['PIK3CA_E4_STA101651']\n",
      "['PIK3CA_E10_STA100188']\n",
      "['PIK3CA_E21_STA100194']\n",
      "['FGFR3_E7_STA101724']\n",
      "['FGFR3_E9_STA100201']\n",
      "['FGFR3_E10_STA101413']\n",
      "['FGFR3_E15_STA101417']\n",
      "['FGFR3_E16_STA101722']\n",
      "['PDGFRA_E11_STA101429']\n",
      "['PDGFRA_E16_STA101432']\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[sample_65]</td>\n",
       "      <td>JAK1_E20_STA100335</td>\n",
       "      <td>5.68358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[sample_65]</td>\n",
       "      <td>JAK1_E18_STA100342</td>\n",
       "      <td>5.68358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[sample_65]</td>\n",
       "      <td>JAK1_E14_STA100350</td>\n",
       "      <td>5.68358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[sample_65]</td>\n",
       "      <td>JAK1_E13_STA100352</td>\n",
       "      <td>5.68358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[sample_65]</td>\n",
       "      <td>JAK1_E5_STA100377</td>\n",
       "      <td>5.68358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461</th>\n",
       "      <td>[sample_96]</td>\n",
       "      <td>FGFR3_E10_STA101413</td>\n",
       "      <td>5.68358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td>[sample_96]</td>\n",
       "      <td>FGFR3_E15_STA101417</td>\n",
       "      <td>5.68358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463</th>\n",
       "      <td>[sample_96]</td>\n",
       "      <td>FGFR3_E16_STA101722</td>\n",
       "      <td>5.68358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464</th>\n",
       "      <td>[sample_96]</td>\n",
       "      <td>PDGFRA_E11_STA101429</td>\n",
       "      <td>5.68358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td>[sample_96]</td>\n",
       "      <td>PDGFRA_E16_STA101432</td>\n",
       "      <td>5.68358</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>466 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            name                  gene    score\n",
       "0    [sample_65]    JAK1_E20_STA100335  5.68358\n",
       "1    [sample_65]    JAK1_E18_STA100342  5.68358\n",
       "2    [sample_65]    JAK1_E14_STA100350  5.68358\n",
       "3    [sample_65]    JAK1_E13_STA100352  5.68358\n",
       "4    [sample_65]     JAK1_E5_STA100377  5.68358\n",
       "..           ...                   ...      ...\n",
       "461  [sample_96]   FGFR3_E10_STA101413  5.68358\n",
       "462  [sample_96]   FGFR3_E15_STA101417  5.68358\n",
       "463  [sample_96]   FGFR3_E16_STA101722  5.68358\n",
       "464  [sample_96]  PDGFRA_E11_STA101429  5.68358\n",
       "465  [sample_96]  PDGFRA_E16_STA101432  5.68358\n",
       "\n",
       "[466 rows x 3 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from tqdm.notebook import tqdm\n",
    "\n",
    "\n",
    "pattern = 'PDGFRA'\n",
    "\n",
    "\n",
    "tp=[]\n",
    "fp=[]\n",
    "tn=[]\n",
    "fn=[]\n",
    "f = pd.DataFrame(columns=[\"name\",\"gene\",\"score\"])\n",
    "q=0\n",
    "for k in tqdm(range(9,10)):\n",
    "    factor = k/10\n",
    "    for p in (range(1)):\n",
    "        synt = pd.read_csv(\"/Users/admin/Documents/CNV/synthetic_raw_85.tsv\",sep=\"\\t\",index_col=0)\n",
    "        synt = synt[np.random.choice(synt.columns,30,replace=False)]\n",
    "        col = np.random.choice(synt.columns,size=1,replace=False)\n",
    "        normalSamples = synt.columns[~np.in1d(synt.columns,col)]\n",
    "\n",
    "        row = np.random.choice(synt.index,size=1)\n",
    "        \n",
    "        tmp = [str(synt.index[i]).split(\"_\")[0] for i in range(synt.shape[0])]\n",
    "        synt.loc[[tmp[i]==pattern for i in range(len(tmp))],col] = synt.loc[[tmp[i]==pattern for i in range(len(tmp))]][col] * factor\n",
    "\n",
    "\n",
    "        neg = synt.index[~np.in1d(tmp,pattern)]\n",
    "        pos = synt.index[np.in1d(tmp,pattern)]\n",
    "\n",
    "        synt_norm = normalizeReads(synt)\n",
    "\n",
    "        conta = 'auto'#1/synt_norm.shape[0]\n",
    "\n",
    "        norm = np.mean(synt_norm[normalSamples], axis = 1)\n",
    "\n",
    "        x = np.array(synt_norm[col]/norm[0])     \n",
    "\n",
    "        random_data = x.reshape(-1, 1)\n",
    "        clf = IsolationForest(contamination=conta).fit(random_data)\n",
    "        preds = clf.predict(random_data)\n",
    "\n",
    "        det = np.array(synt_norm.index)[preds==-1]\n",
    "        ndet = np.array(synt_norm.index)[preds==1]\n",
    "               \n",
    "        for gene in det:\n",
    "            r = re.compile(gene)\n",
    "            abEx = list(filter(r.match, det))\n",
    "            tmp = synt.filter(regex=\"^\"+gene,axis=0)                \n",
    "            score = scoreAmplif(len(abEx),tmp.shape[0],synt.shape[0])\n",
    "\n",
    "            f.loc[q] = [col,gene,score]\n",
    "            q=q+1\n",
    "\n",
    "                \n",
    "f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9\n"
     ]
    }
   ],
   "source": [
    "for k in range(9,10):\n",
    "    print(k)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
